src/builder.ts

```diff
@@ -486,9 +513,18 @@ export class SchemaBuilder {
     }
   }
 
-  getFinalTypeMap(): BuildTypes<any> {
-    this.finalized = true;
-    this.walkTypes();
+  beforeWalkTypes() {
+    this.plugins.forEach(({ config: pluginConfig }) => {
+      if (pluginConfig.onInstall) {
+        pluginConfig.onInstall(this.builderLens);
+      }
+      if (pluginConfig.onCreateFieldResolver) {
+        this.onCreateResolverFns.push(pluginConfig.onCreateFieldResolver);
+      }
+    });
+  }
```

Comment:
We'll probably discussion this on my PR but a quite different approach I took with plugins was colocating the lifecycle logic and interface and having only a very limited `Controller` concept that the rest of the system could use to trigger.

Can't easily tell which approach is better. Yours has one level less of indirection but a weaker encapsulation / representation of plugin. For example the nullability of `onInstall` is leaked and made a concern outside of the plugin module. The `builderLens` is owned by builder rather than plugin, but arguably a the lens is an external concept to the builder, and introduced by the plugin component, completely independent of the builder. But to counter myself, if other parts of the system come to desire a `lens` variant of builder, then your system is arguably better positioned to serve that need (as why would plugin module be a better distributor of the builder lens than the builder). But that's hypothetical AFAIK, lens is really a concept driven by plugins (state property above for lens on this class says as much). 

You've looked at a broader surface area of the plugins in this PR and surely saw couplings I didn't have to deal with, like `onCreateResolverFns`.

Interesting stuff anyways.